internal/cmd/stack/run_list.go

```diff
@@ -65,37 +67,86 @@ type runsJSONQuery struct {
 	TriggeredBy    string `graphql:"triggeredBy" json:"triggeredBy"`
 }
 
-func listRunsJSON(ctx context.Context, stackID string, maxResults int) error {
-	var results []runsJSONQuery
+func (r runsJSONQuery) Cursor() string {
+	return r.ID
+}
+
+type withCursor interface {
+	Cursor() string
+}
+
+func queryTrackedRuns[T any](ctx context.Context, stackID string, before *string) ([]T, error) {
+	var query struct {
+		Stack *struct {
+			Runs []T `graphql:"runs(before: $before)"`
+		} `graphql:"stack(id: $stackId)"`
+	}
+
+	if err := authenticated.Client.Query(ctx, &query, map[string]interface{}{"stackId": stackID, "before": before}); err != nil {
+		return nil, errors.Wrap(err, "failed to query run list")
+	}
+
+	if query.Stack == nil {
+		return nil, fmt.Errorf("stack %q not found", stackID)
+	}
+
+	return query.Stack.Runs, nil
+}
+
+func queryPreviewRuns[T any](ctx context.Context, stackID string, before *string) ([]T, error) {
+	var query struct {
+		Stack *struct {
+			ProposedRuns []T `graphql:"proposedRuns(before: $before)"`
+		} `graphql:"stack(id: $stackId)"`
+	}
+
+	if err := authenticated.Client.Query(ctx, &query, map[string]interface{}{"stackId": stackID, "before": before}); err != nil {
+		return nil, errors.Wrap(err, "failed to query run list")
+	}
+
+	if query.Stack == nil {
+		return nil, fmt.Errorf("stack %q not found", stackID)
+	}
+
+	return query.Stack.ProposedRuns, nil
+
+}
+
+func fetchRuns[T withCursor](ctx context.Context, stackID string, maxResults int, preview bool) ([]T, error) {
+	var results []T
 	var before *string
 
 	for len(results) < maxResults {
-		var query struct {
-			Stack *struct {
-				Runs []runsJSONQuery `graphql:"runs(before: $before)"`
-			} `graphql:"stack(id: $stackId)"`
+		var err error
+		var runs []T
+		if preview {
```

Comment:
Could we pass a "fetcher" function here rather than a bool? We could decide which function to pass in the `runList` function.

It would make the `fetchRuns` and `listRuns(Table,JSON)` functions easier to test.

Reply:
I like that. I'll change it in a moment